pkg/comp-functions/functions/vshnmariadb/mariadb_deploy.go

```diff
@@ -0,0 +1,263 @@
+
+	l := controllerruntime.LoggerFrom(ctx)
+
+	comp := &vshnv1.VSHNMariaDB{}
+	err := svc.GetObservedComposite(comp)
+	if err != nil {
+		err = fmt.Errorf("cannot get observed composite: %w", err)
+		return runtime.NewFatalResult(err)
+	}
+
+	l.Info("Creating namespace for mariadb instance")
+	err = createObjectNamespace(ctx, comp, svc)
+	if err != nil {
+		err = fmt.Errorf("cannot create mariadb namespace: %w", err)
+		return runtime.NewFatalResult(err)
+	}
+
+	l.Info("Creating tls certificate for mariadb instance")
+	err = common.CreateTlsCerts(ctx, comp.GetInstanceNamespace(), comp.GetName(), svc)
+	if err != nil {
+		err = fmt.Errorf("cannot create tls certificate: %w", err)
+		return runtime.NewFatalResult(err)
+	}
+
+	l.Info("Creating helm release for mariadb instance")
+	err = createObjectHelmRelease(ctx, comp, svc)
+	if err != nil {
+		err = fmt.Errorf("cannot create helm release: %w", err)
+		return runtime.NewFatalResult(err)
+	}
+
+	l.Info("Creating network policies mariadb instance")
+	sourceNs := []string{comp.GetClaimNamespace()}
+	if svc.GetBoolFromCompositionConfig("slosEnabled") {
+		sourceNs = append(sourceNs, svc.Config.Data["slosNs"])
+	}
+	err = common.CreateNetworkPolicy(ctx, sourceNs, comp.GetInstanceNamespace(), comp.GetName(), svc)
+	if err != nil {
+		err = fmt.Errorf("cannot create helm release: %w", err)
+		return runtime.NewFatalResult(err)
+	}
+
+	l.Info("Get connection details from secret")
+	err = getConnectionDetails(ctx, comp, svc)
+	if err != nil {
+		if err == runtime.ErrNotFound {
+			return runtime.NewNormalResult("skipping sli bucket, connectiondetails not yet available")
+		}
+		err = fmt.Errorf("cannot get connection details: %w", err)
+		return runtime.NewFatalResult(fmt.Errorf("cannot get connection details: %w", err))
+	}
+	return nil
+}
+
+// Create the namespace for the mariadb instance
+func createObjectNamespace(ctx context.Context, comp *vshnv1.VSHNMariaDB, svc *runtime.ServiceRuntime) error {
+
+	ns := &corev1.Namespace{
+
+		ObjectMeta: metav1.ObjectMeta{
+			Name: comp.GetInstanceNamespace(),
+			Labels: map[string]string{
+				"appcat.vshn.io/servicename":     "mariadb-standalone",
+				"appcat.vshn.io/claim-namespace": comp.GetClaimNamespace(),
+				"appuio.io/no-rbac-creation":     "true",
+				"appuio.io/billing-name":         "appcat-mariadb"},
+		},
+	}
+
+	return svc.SetDesiredKubeObject(ns, comp.Name+"-ns")
+}
+
+// Create the helm release for the mariadb instance
+func createObjectHelmRelease(ctx context.Context, comp *vshnv1.VSHNMariaDB, svc *runtime.ServiceRuntime) error {
+
+	plan := comp.Spec.Parameters.Size.GetPlan(svc.Config.Data["defaultPlan"])
+
+	resouces, err := utils.FetchPlansFromConfig(ctx, svc, plan)
+	if err != nil {
+		err = fmt.Errorf("cannot fetch plans from the composition config, maybe they are not set: %w", err)
+		return err
+	}
+
+	reqMem := comp.Spec.Parameters.Size.Requests.Memory
```

Comment:
NIT: This looks like something we could generalize, too.

Reply:
Extracted this part to a new generic function